src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1075,8 +1080,47 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
             }
 
             diagnostics.AddRangeAndFree(localDiagnostics);
-            var boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, type: declTypeOpt.TypeSymbol);
+
+            BoundTypeExpression boundDeclType = null;
+
+            if (bindType)
+            {
+                var invalidDimensions = ArrayBuilder<BoundExpression>.GetInstance();
+
+                // We don't want to report any errors from this, as the syntax is invalid anyway
+                var dummyDiagnostics = DiagnosticBag.GetInstance();
+
+                GetInvalidDimensions(typeSyntax, invalidDimensions, dummyDiagnostics);
+
+                dummyDiagnostics.Free();
+
+                boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, dimensionsOpt: invalidDimensions.ToImmutableAndFree(), type: declTypeOpt.TypeSymbol);
+            }
+
             return new BoundLocalDeclaration(associatedSyntaxNode, localSymbol, boundDeclType, initializerOpt, arguments, hasErrors);
+
+            void GetInvalidDimensions(TypeSyntax node, ArrayBuilder<BoundExpression> invalidDimensions, DiagnosticBag diagnosticsBag)
+            {
+                foreach (var child in node.ChildNodesAndTokens())
+                {
+                    if (child.AsNode() is TypeSyntax childNode)
+                    {
+                        GetInvalidDimensions(childNode, invalidDimensions, diagnosticsBag);
```

Comment:
>GetInvalidDimensions(childNode, invalidDimensions, diagnosticsBag); [](start = 24, length = 67)

It is hard to be confident that this is never going to descent into the array sizes and, if it is going to descent into them, then it won't cause any duplicates added into the ```invalidDimensions```. Consider restructuring the code in the way that it is clear what children of ArrayTypeSyntax we visit, when and how.

Reply:
Would you advise inheriting syntax visitor for this?

Reply:
> Would you advise inheriting syntax visitor for this?

I would use a syntax visitor, but simply not going into the ```foreach (var child in node.ChildNodesAndTokens())``` when we are dealing with ```ArrayTypeSyntax``` and instead explicitly handling its children (you already handle sizes, what is left is to handle element type) should work too.

---
In reply to: [255648924](https://github.com/dotnet/roslyn/pull/32843#discussion_r255648924) [](ancestors = 255648924)